plugin/action/throttle/redis_limiter.go

```diff
@@ -0,0 +1,185 @@
+	redis redisClient
+
+	// <keyPrefix>_
+	// bucket counter prefix bucket id forms key in redis: <keyPrefix>_<bucketID>
+	keyPrefix bytes.Buffer
+
+	// <keyPrefix>_<keySuffix>
+	// limit key in redis
+	keyLimit string
+
+	// contains values which will be used for incrementing remote bucket counter
+	// buckets will be flushed after every sync to contain only increment value
+	incrementLimiter *inMemoryLimiter
+
+	// contains global values synced from redis
+	totalLimiter *inMemoryLimiter
+
+	// contains indexes of buckets in incrementLimiter for sync
+	keyIdxsForSync []int
+
+	// contains bucket IDs for keys in keyIdxsForSync
+	bucketIdsForSync []int
+}
+
+// NewRedisLimiter return instance of redis limiter.
+func NewRedisLimiter(
+	ctx context.Context,
+	redis redisClient,
+	pipelineName, throttleFieldName, throttleFieldValue string,
+	bucketInterval time.Duration,
+	bucketCount int,
+	limit complexLimit,
+) *redisLimiter {
+	rl := &redisLimiter{
+		redis:            redis,
+		incrementLimiter: NewInMemoryLimiter(bucketInterval, bucketCount, limit),
+		totalLimiter:     NewInMemoryLimiter(bucketInterval, bucketCount, limit),
+	}
+
+	rl.keyIdxsForSync = make([]int, 0, bucketCount)
+	rl.bucketIdsForSync = make([]int, 0, bucketCount)
+	rl.keyPrefix = bytes.Buffer{}
+
+	// full name of keyPrefix will be pipelineName_throttleFieldName_throttleFieldValue_limit. `limit` added afterwards
+	rl.keyPrefix.WriteString(pipelineName)
+	rl.keyPrefix.WriteString("_")
+	rl.keyPrefix.WriteString(throttleFieldName)
+	rl.keyPrefix.WriteString("_")
+	rl.keyPrefix.WriteString(throttleFieldValue)
+	rl.keyPrefix.WriteString("_")
+	rl.keyLimit = rl.keyPrefix.String() + keySuffix
+
+	return rl
+}
+
+func (l *redisLimiter) isAllowed(event *pipeline.Event, ts time.Time) bool {
+	// count increment from last sync
+	isAllowed := l.incrementLimiter.isAllowed(event, ts)
+	if isAllowed {
+		// count global limiter from last sync
+		isAllowed = l.totalLimiter.isAllowed(event, ts)
+	}
+
+	return isAllowed
+}
+
+func (l *redisLimiter) sync() {
+	// required to prevent concurrent update of buckets via throttle plugin
+	l.incrementLimiter.mu.Lock()
+	l.totalLimiter.mu.Lock()
+
+	n := time.Now()
+
+	// actualize buckets
+	maxID := l.incrementLimiter.rebuildBuckets(n)
+	_ = l.totalLimiter.rebuildBuckets(n)
+
+	minID := l.totalLimiter.minID
+	count := l.incrementLimiter.bucketCount
+
+	l.keyIdxsForSync = l.keyIdxsForSync[:0]
+	l.bucketIdsForSync = l.bucketIdsForSync[:0]
+	for i := 0; i < count; i++ {
+		// no new events passed
+		if l.incrementLimiter.buckets[i] == 0 {
+			continue
+		}
+		l.keyIdxsForSync = append(l.keyIdxsForSync, i)
+		l.bucketIdsForSync = append(l.bucketIdsForSync, minID+i)
+	}
+
+	l.totalLimiter.mu.Unlock()
+	l.incrementLimiter.mu.Unlock()
+
+	l.syncLocalGlobalLimiters(maxID)
+	l.updateKeyLimit()
+}
+
+func (l *redisLimiter) syncLocalGlobalLimiters(maxID int) {
+	prefix := l.keyPrefix.String()
+
+	for i, ID := range l.bucketIdsForSync {
+		i := i
+		ID := ID
```

Comment:
why you need this?

Reply:
It remains from previous optimizations and it stale now. Removed.